aten/src/ATen/core/op_registration/base.h

```diff
@@ -0,0 +1,81 @@
+#pragma once
+
+/**
+ * This file sets up the basics for operator registration like the
+ * c10::RegisterOperators() class.
+ *
+ * You probably don't want to include this file directly but include
+ * op_registration.h instead since that adds more functionality you'll
+ * likely need to register your operators.
+ */
+
+#include <ATen/core/dispatch/Dispatcher.h>
+
+namespace c10 {
+
+namespace detail {
+
+  // OperatorRegistrar in its constructor registers an operator in the dispatch
+  // table deregisters it in the destructor. The intent is that this class is
+  // constructed at static initialization time so that operators automatically
+  // get registered when a dlopen() occurs.
+  // You shouldn't call this directly; instead, use the RegisterOperators class.
+  class OperatorRegistrar final {
+  public:
+    explicit OperatorRegistrar(FunctionSchema&& schema, TensorTypeId dispatch_key, KernelFunction* kernel, KernelCacheCreatorFunction* cache_creator)
+    : op_(Dispatcher::singleton().registerSchema(std::move(schema))), dispatch_key_(std::move(dispatch_key)), owns_registration_(true) {
+      Dispatcher::singleton().registerKernel(op_, dispatch_key_, kernel, cache_creator);
+    }
+
+    OperatorRegistrar(OperatorRegistrar&& rhs) noexcept
+    :  op_(std::move(rhs.op_)), dispatch_key_(std::move(rhs.dispatch_key_)), owns_registration_(rhs.owns_registration_) {
+      rhs.owns_registration_ = false;
+    }
+
+    // not needed and would break RAII if defaulted.
+    OperatorRegistrar& operator=(OperatorRegistrar&& rhs) noexcept = delete;
+    OperatorRegistrar(const OperatorRegistrar& rhs) = delete;
+    OperatorRegistrar& operator=(const OperatorRegistrar& rhs) = delete;
+
+    ~OperatorRegistrar() {
+      if (owns_registration_) {
+        Dispatcher::singleton().deregisterKernel(op_, dispatch_key_);
+        Dispatcher::singleton().deregisterSchema(op_);
+      }
+    }
+
+  private:
+    const OperatorHandle op_;
+    const TensorTypeId dispatch_key_;
+    bool owns_registration_;
```

Comment:
nit: is it only to make move constructor work? can't we just set the op_ to some bad value? they don't technically need to be const, do they?

Reply:
yes, it's just for the move constructor. An alternative would be to have std::optional<OperatorHandle>. That wouldn't give us any size though (since optional just has the bool internally). Also, it's kind of weird because then still the dispatch_key_ would be there. So std::optional<RegistrationData> with struct RegistrationData containing OperatorHandle and TensorTypeId? Seems unnecessarily complex. I think the `bool owns_registration_` is the simplest solution.